src/main/java/org/springframework/data/neo4j/core/ReactiveNeo4jTemplate.java

```diff
@@ -562,140 +566,224 @@ private Object convertIdValues(@Nullable Neo4jPersistentProperty idProperty, Obj
+			Neo4jPersistentProperty relationshipProperty = association.getInverse();
+
+			stateMachine.markAsProcessed(relationshipDescription, relatedValuesToStore);
+			Flux<RelationshipPropertyValues> relationshipCreation = Flux.fromIterable(relatedValuesToStore).flatMap(relatedValueToStore -> {
+
+				Object relatedNodePreEvt = relationshipContext.identifyAndExtractRelationshipTargetNode(relatedValueToStore);
+				return Mono.deferContextual(ctx -> eventSupport.maybeCallBeforeBind(relatedNodePreEvt)
+						.flatMap(relatedNode -> Mono.just(neo4jMappingContext.getPersistentEntity(relatedNodePreEvt.getClass()))
+								.flatMap(targetEntity -> Mono.just(targetEntity.isNew(relatedNode))
+										.flatMap(isNew -> saveRelatedNode(relatedNode, relationshipContext.getAssociationTargetType(),
+												targetEntity)
+												.flatMap(relatedInternalId -> {
+													// if an internal id is used this must get set to link this entity in the next iteration
+													PersistentPropertyAccessor<?> targetPropertyAccessor = targetEntity
+															.getPropertyAccessor(relatedNode);
+													if (targetEntity.isUsingInternalIds()) {
+														targetPropertyAccessor.setProperty(targetEntity.getRequiredIdProperty(),
+																relatedInternalId);
+													}
+
+													CreateRelationshipStatementHolder statementHolder = neo4jMappingContext.createStatement(
+															sourceEntity, relationshipContext, relatedValueToStore);
+
+													// in case of no properties the bind will just return an empty map
+													return neo4jClient
+															.query(renderer.render(statementHolder.getStatement()))
+															.bind(convertIdValues(sourceEntity.getRequiredIdProperty(), fromId)) //
+															.to(Constants.FROM_ID_PARAMETER_NAME) //
+															.bind(relatedInternalId) //
+															.to(Constants.TO_ID_PARAMETER_NAME) //
+															.bindAll(statementHolder.getProperties())
+															.fetchAs(Long.class).one()
+															.flatMap(relationshipInternalId -> {
+																if (idProperty != null) {
+																	relationshipContext
+																			.getRelationshipPropertiesPropertyAccessor(relatedValueToStore)
+																			.setProperty(idProperty, relationshipInternalId);
+																}
+
+																Mono<Object> something = null;
+																if (processState != ProcessState.PROCESSED_ALL_VALUES) {
+																	something = processNestedRelations(targetEntity, targetPropertyAccessor,
+																			isNew, stateMachine);
+																} else {
+																	something = Mono.just((T) targetPropertyAccessor.getBean());
+																}
+
+																return something.flatMap(currentValue -> {
+																	Object newRelationshipObject = targetPropertyAccessor.getBean();
+																	if (relationshipDescription.hasRelationshipProperties()) {
+																		Object lalelu = relationshipProperty.isDynamicAssociation()
```

Comment:
I would rather prefer to bite it and deduplicate this conditions, similar on what we did together with the checks for cyclic dependencies, regardless of the fact that one lives in the imperative world and the other in the reactive world.